Sources/Testing/SourceAttribution/CustomTestStringConvertible.swift

```diff
@@ -0,0 +1,154 @@
+/// `String(describing:)`. The resulting string may be inappropriate for some
+/// types and their values. If the type of the value is made to conform to
+/// ``CustomTestStringConvertible``, then the value of its ``testDescription``
+/// property will be used instead.
+///
+/// For example, consider the following type:
+///
+/// ```swift
+/// enum Food {
+///   case paella
+///   case oden
+///   case ragu
+///   ...
+/// }
+/// ```
+///
+/// If an array of cases from this enumeration is passed to a parameterized test
+/// function:
+///
+/// ```swift
+/// @Test(arguments: [.paella, .oden, .ragu])
+/// func isDelicious(_ food: Food) { ... }
+/// ```
+///
+/// Then the values in the array need to be presented in the test output, but
+/// the default description of a value may not be adequately descriptive:
+///
+/// ```
+/// ◇ Passing argument food → .paella to isDelicious(\_:)
+/// ◇ Passing argument food → .oden to isDelicious(\_:)
+/// ◇ Passing argument food → .ragu to isDelicious(\_:)
+/// ```
+///
+/// By adopting ``CustomTestStringConvertible``, customized descriptions can be
+/// included:
+///
+/// ```swift
+/// extension Food: CustomTestStringConvertible {
+///   var testDescription: String {
+///     switch self {
+///     case .paella:
+///       "paella valenciana"
+///     case .oden:
+///       "おでん"
+///     case .ragu:
+///       "ragù alla bolognese"
+///     }
+///   }
+/// }
+/// ```
+///
+/// The presentation of these values will then reflect the value of the
+/// ``testDescription`` property:
+///
+/// ```
+/// ◇ Passing argument food → paella valenciana to isDelicious(\_:)
+/// ◇ Passing argument food → おでん to isDelicious(\_:)
+/// ◇ Passing argument food → ragù alla bolognese to isDelicious(\_:)
+/// ```
+///
+/// ## See Also
+///
+/// - ``Swift/String/init(describingForTest:)``
+public protocol CustomTestStringConvertible {
+  /// A description of this instance to use when presenting it in a test's
+  /// output.
+  ///
+  /// Do not use this property directly. To get the test description of a value,
+  /// use ``Swift/String/init(describingForTest:)``.
+  var testDescription: String { get }
+}
+
+extension String {
+  /// Initialize this instance so that it can be presented in a test's output.
+  ///
+  /// - Parameters:
+  ///   - value: The value to describe.
+  ///
+  /// ## See Also
+  ///
+  /// - ``CustomTestStringConvertible``
+  public init<T>(describingForTest value: T) {
+    // The mangled type name SPI doesn't handle generic types very well, so we
+    // ask for the dynamic type of `value` (type(of:)) instead of just T.self.
+    lazy var valueType = type(of: value as Any)
+    if let value = value as? any CustomTestStringConvertible {
+      self = value.testDescription
+    } else if let value = value as? any CustomStringConvertible {
+      self.init(describing: value)
+    } else if let value = value as? any CustomDebugStringConvertible {
+      self.init(reflecting: value)
```

Comment:
I did think of that, but I left it out mostly because I don't think it's a very common protocol to conform to… but it's trivial to add it. Strong feelings here?

Reply:
Actually, could these 3 known stdlib protocol cases be collapsed into a single `else if` clause, since `String(describing:)` is known to handle them?
```swift
// If it's one of the known types that `String(describing:)` prioritizes, defer to that
} else if value is TextOutputStreamable
       || value is CustomStringConvertible
       || value is CustomDebugStringConvertible {
  self.init(describing: value)
} else if /* C enum */ {
 ...
} else if /* Swift enum */ {
 ...
} else {
  // Use the generic description of the value.
  self.init(describing: value)
}
```

Reply:
Yeah, we can do that.

Reply:
Actually, no, let's not, because we lose access to the specializations on `String.init(describing:)` and that means the runtime will end up needing to do twice as many conformance checks, which isn't free. See: https://github.com/apple/swift/blob/06984b430775bd3ae76d705a741d2427e6166fe1/stdlib/public/core/Mirror.swift#L625

Reply:
Ah right, makes sense. Well, I think I'd lean towards adding the `TextOutputStreamable` specialization at least